src/core/geometry/qgslinestring.cpp

```diff
@@ -326,17 +326,104 @@ bool QgsLineString::isValid( QString &error, Qgis::GeometryValidityFlags flags )
   return QgsCurve::isValid( error, flags );
 }
 
-QgsLineString *QgsLineString::snappedToGrid( double hSpacing, double vSpacing, double dSpacing, double mSpacing ) const
+QgsLineString *QgsLineString::snappedToGrid( double hSpacing, double vSpacing, double dSpacing, double mSpacing, bool removeRedundantPoints ) const
 {
-  // prepare result
-  std::unique_ptr<QgsLineString> result { createEmptyWithSameType() };
+  const int length = numPoints();
+  if ( length < 2 )
+    return nullptr;
 
-  bool res = snapToGridPrivate( hSpacing, vSpacing, dSpacing, mSpacing, mX, mY, mZ, mM,
-                                result->mX, result->mY, result->mZ, result->mM );
-  if ( res )
-    return result.release();
-  else
+  const bool hasZ = is3D();
+  const bool hasM = isMeasure();
+
+  QVector< double > outX;
+  outX.reserve( length );
+  QVector< double > outY;
+  outY.reserve( length );
+  QVector< double > outZ;
+  if ( hasZ )
+    outZ.reserve( length );
+  QVector< double > outM;
+  if ( hasM )
+    outM.reserve( length );
+
+  const double *xIn = mX.constData();
+  const double *yIn = mY.constData();
+  const double *zIn = hasZ ? mZ.constData() : nullptr;
+  const double *mIn = hasM ? mM.constData() : nullptr;
+
+  double previousX = 0;
+  double previousY = 0;
+  double previousZ = 0;
+  double previousM = 0;
+  int outSize = 0;
+  for ( int i = 0; i < length; ++i )
+  {
+    const double currentX = *xIn++;
+    const double currentY = *yIn++;
+    const double currentZ = zIn ? *zIn++ : 0;
+    const double currentM = mIn ? *mIn++ : 0;
+
+    const double roundedX = hSpacing > 0 ? ( std::round( currentX / hSpacing ) * hSpacing ) : currentX;
+    const double roundedY = vSpacing > 0 ? ( std::round( currentY / vSpacing ) * vSpacing ) : currentY;
+    const double roundedZ = hasZ && dSpacing > 0 ? ( std::round( currentZ / dSpacing ) * dSpacing ) : currentZ;
+    const double roundedM = hasM && mSpacing > 0 ? ( std::round( currentM / mSpacing ) * mSpacing ) : currentM;
+
+    if ( i > 0 )
+    {
+      const bool isPointEqual = qgsDoubleNear( roundedX, previousX )
+                                && qgsDoubleNear( roundedY, previousY )
+                                && ( !hasZ || dSpacing <= 0 || qgsDoubleNear( roundedZ, previousZ ) )
+                                && ( !hasM || mSpacing <= 0 || qgsDoubleNear( roundedM, previousM ) );
+      if ( isPointEqual )
+        continue;
+
+      // maybe previous point is redundant and is just a midpoint on a straight line -- let's check
+      bool previousPointRedundant = false;
+      if ( outSize > 1 && !hasZ && !hasM )
```

Comment:
Should we not be removing points that end up having the same `x,y` coordinates but different `z` or `m` as redundant?
I'd even argue that we should be removing those duplicates even with `removeReduntantPoints == false`!

Reply:
> Should we not be removing points that end up having the same x,y coordinates but different z or m as redundant?
I'd even argue that we should be removing those duplicates even with removeReduntantPoints == false!

I don't think so. At least, I'd argue that we should only do that if the z/m value of the point would be the same if it were linearly interpolated from the vertices left intact at the original redundant point x/y. And that logic would be non-trivial and isn't relevant for my pressing use case (https://github.com/qgis/QGIS/pull/57623).